### PR TITLE
BasicIdToken files

### PR DESCRIPTION
**Pull request by anjuthomas**

### CODE REVIEW

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
+'use strict';
+const Message = require('../message');
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js:2]: Where is the Message class defined?

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+const Token = require('./token');
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js:3]: Where is token defined?

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js:2]: Just curious, if you are using ES6, why not use ES6 `import` instead of `require`:
`import * as Message from '../message';`

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ * Optional claims : aud, exp, nbf
+ */
+
+/**
+ * BasicIdToken
+ * Init token using required claims
+ * @class
+ * @constructor
+ * @extends Message
+ * @param {*} iss
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: The constructor seems to accept one parameter object `{iss, sub, iat, jti}`. The javadoc here is not correct.
Also shouldn't all these parameter types by `string` instead of `*`?

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      'notBefore',
+      'noTimestamp',
+      'audience',
+      'issuer',
+      'subject',
+      'jwtid',
+    ];
+
+    /** Known required claims */
+    this.knownOptionalClaims = {
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: This property and the ones below it seem like they could be defined as an array like `this.optionsForObjects`.
`forEach` can also be used to iterate over an array.
Is this the format as defined in the `Message` class?

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    this.optionsToPayload = {
+      'iss': 'iss',
+      'sub': 'sub',
+      'iat': 'iat',
+      'jti': 'jti',
+    };
+  }
+
+
+  static init(payload, options) {
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: Can you provide javadoc for this

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    });
+    basicIdToken.addOptionalClaims(optionalClaims);
+    if (options && Object.keys(options).indexOf('algorithm') !== -1 &&
+        options['algorithm'] === 'none') {
+      basicIdToken.setNoneAlgorithmAttr(true);
+    }
+    return basicIdToken;
+  }
+
+  static toJWT(payload, key, options) {
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: Can you provide javadoc for this

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    let basicIdToken = this.init(payload, options);
+    return basicIdToken.toJWT(key, options);
+  }
+
+  static fromJWT(jwt, key, verificationClaims, options) {
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: Can you provide javadoc for this

File: test/basicId-claims-test.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+var BasicIdToken = require('../src/oicMsg/tokenProfiles/basicIdToken');
+
+function loadKey(filename) {
+  return fs.readFileSync(path.join(__dirname, filename));
+}
+
+var algorithms = {
+  RS256: {
+    pub_key: loadKey('pub.pem'),
```
Comment (bojeil-google) [inline test/basicId-claims-test.js]: Where are these test key files located? I don't see them in the repo.
Do you expect them to be generated before the tests are run?

File: test/basicId-claims-test.js [hunk 9]
```diff
@@ -0,0 +1,226 @@
+var fs = require('fs');
```
Comment (bojeil-google) [inline test/basicId-claims-test.js]: You seem to be using ES5 for the unit tests. Why not stick to ES6?

File: test/basicId-claims-test.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+var path = require('path');
+var assert = require('chai').assert;
```
Comment (bojeil-google) [inline test/basicId-claims-test.js]: consider using `chai.expect`.
It is easier to assert if some call throws an error without having to try/catch.
eg:
```
expect(() => {
  failingCallback();
}).to.throw('Some error');
```

File: test/basicId-claims-test.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+                        'iss': 'issuer',
+                        'sub': 'subject',
+                        'aud': 'audience',
+                        'maxAge': '1d',
+                        'clockTolerance': 10,
+                        'jti': 'jti'
+                      },
+                      {'clockTimestamp': clockTimestamp})
+                } catch (err) {
+                  assert.isNull(err);
```
Comment (bojeil-google) [inline test/basicId-claims-test.js:60]: Are you expecting this to be called? It is not clear what the expectation here is. I would expect that you are testing that the function call succeeds. 
A cleaner way is to use `expect().to.throw()` here and below

File: test/basicId-claims-test.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+                      'jti': 'jti'
+                    },
+                    {'clockTimestamp': clockTimestamp})
+                .then(function(decodedPayload) {
+                  assert.isNotNull(decodedPayload);
+                })
+                .catch(function(err) {
+                  assert.isNull(err);
+                });
+            done();
```
Comment (bojeil-google) [inline test/basicId-claims-test.js]: For asynchronous tests, `done` should be called when the promise completes. You are calling it prematurely ehre and elsewhere.

File: src/oicMsg/tokenProfiles/basicIdToken.js [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    const basicIdToken = new BasicIdToken(payload);
+    let optionalClaims = {};
+    Object.keys(basicIdToken.knownOptionalClaims).forEach(key => {
+      if (payload[key]) {
+        optionalClaims[key] = payload[key];
+      }
+    });
+    basicIdToken.addOptionalClaims(optionalClaims);
+    if (options && Object.keys(options).indexOf('algorithm') !== -1 &&
+        options['algorithm'] === 'none') {
```
Comment (bojeil-google) [inline src/oicMsg/tokenProfiles/basicIdToken.js]: Can you also add a unit test for this to confirm algorithm is set to none automatically

**Review by bojeil-google — CHANGES_REQUESTED** [review]

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: It has not been reviewed it. Since you requested to have only one PR at a time I thought I 
would start with BasicIdToken.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 2 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:3]: It has not been reviewed it. Since you requested to have only one PR at a time I thought I
would start with BasicIdToken.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: Good point. Will have to change this in a lot of places through out the library. I can make the change, but is this needed?

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 3 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: Fixed

File: test/basicId-claims-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Answered above.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 3 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: Fixed

File: test/basicId-claims-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Fixed

File: test/basicId-claims-test.js (hunk 12 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Fixed

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 3 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: Fixed

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 4 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: Fixed

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (bojeil-google) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: Hmm, we should have started with `Message` or `Token` since this depends on it. It's hard for me to review without knowing anything about these classes. Also your tests will fail and nothing will compile when you merge this PR. Would it be possible to send parallel PRs for these 2 classes.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 2 above)
Comment (bojeil-google) [reply src/oicMsg/tokenProfiles/basicIdToken.js:3]: Send me the PR for `Token` to review in parallel.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (bojeil-google) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: It would make sense to be consistent. You already implemented this in es6.
It would also make sense to do that early in the process. It will get harder later on. You can keep it as is for now and do it in a follow up.

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 3 above)
Comment (bojeil-google) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: I don't see your change. Did you forget to commit it?

File: test/basicId-claims-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Fixed

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 3 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js]: I haven't pushed the changes yet. Will update you once I push them

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 2 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:3]: Done

File: src/oicMsg/tokenProfiles/basicIdToken.js (hunk 1 above)
Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: Done

Comment (anjuthomas) [reply src/oicMsg/tokenProfiles/basicIdToken.js:2]: Sounds good. On my todo list as need  to add additional setting up to enable this.

File: test/basicId-claims-test.js (hunk 12 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Fixed. Will push the commit after the other tasks are complete

File: test/basicId-claims-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: Fixed. Will push the commit after the other tasks are complete

File: test/basicId-claims-test.js (hunk 10 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: On my todo list

File: test/basicId-claims-test.js (hunk 11 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: On my todo list

File: test/basicId-claims-test.js (hunk 12 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js]: On my todo list

File: test/basicId-claims-test.js (hunk 11 above)
Comment (anjuthomas) [reply test/basicId-claims-test.js:60]: expect does not seem to work well with async functions compared to asserts
